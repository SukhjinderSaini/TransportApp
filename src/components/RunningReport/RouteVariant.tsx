import React from "react";

interface Props {
  formatString: string;
}

function RouteVariant({ formatString }: Props) {
  if (formatString === "UNKNOWN" || formatString.length < 3) {
    return <span>{formatString}</span>;
  }
  const charactersToFormat = formatString.slice(0, 3);
  return (
    <>
      <span className="fw-bold">{charactersToFormat}</span>
      <span>{formatString.substring(3)}</span>
    </>
  );
}

export default RouteVariant;
