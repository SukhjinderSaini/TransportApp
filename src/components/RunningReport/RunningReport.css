.organisation{
    width:200px;
    min-width: 50px;
    height: 200px;
    border-radius: 50%;
    border: 2px dashed;
    margin: 30px;
    cursor: pointer;
}
.caret_border{
    border:4px solid #000;
    border-radius: 7px;
}
th{
    background: #e3f2fd !important;
}
.accordion-button::after{
    display:none;
}
.accordion-button {
  padding:9px 0px;
  background-color: unset !important;
}

.accordion-button:focus {
  box-shadow: none;
}

.accordion-button:not(.collapsed) {
  color: #212529;
}
.accordion-body{
    padding:0;
}