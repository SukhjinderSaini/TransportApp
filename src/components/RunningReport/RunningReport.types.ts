type RunningDateType = string;

export interface BusRunningStatus {
  organisation: string;
  date: RunningDateType;
  busData: BusRunningInfo[];
}

export interface BusRunningInfo {
  busId: string;
  routeVariant: string;
  deviationFromTimetable: number;
}
