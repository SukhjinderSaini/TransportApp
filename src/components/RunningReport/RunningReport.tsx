import React, { useState, useEffect } from "react";
import "./RunningReport.css";
import { BusRunningStatus, BusRunningInfo } from "./RunningReport.types";
import RouteVariant from "./RouteVariant";
import ServiceNotes from "../ServiceNotes";
import { Table, Accordion } from "react-bootstrap";
import { FaRegCaretSquareDown, FaRegCaretSquareUp } from "react-icons/fa";
import { BsCaretUpFill, BsCaretDownFill } from "react-icons/bs";

function RunningReport() {
  const [busData, setBusData] = useState<BusRunningStatus[]>();
  const [activeKey, setActiveKey] = useState<string | null>(null);
  const fetchServiceData = async () => {
    const bus_service_call = await fetch("/data/bus-services-data.json");
    const bus_service_data = await bus_service_call.json();
    return bus_service_data.data;
  };
  useEffect(() => {
    fetchServiceData().then((data) => {
      setBusData(data);
    });
  }, []);
  const handleAccordionClick = (index: string) => {
    const updateIndex = activeKey === index ? null : index;
    setActiveKey(updateIndex);
  };
  return (
    <>
      <div className="d-flex flex-column m-sm-3 m-md-5">
        <div className="display-6 fw-normal text-start">Bus Reports</div>
        <div className="aling-items-center justify-content-center mt-3">
          {busData &&
            Array.isArray(busData) &&
            busData.length > 0 &&
            busData.map((serviceData: BusRunningStatus, index) => (
              <Accordion
                flush
                className="ma-5"
                key={serviceData.organisation + "_" + index}
                defaultActiveKey="1"
                activeKey={activeKey}>
                <Accordion.Item eventKey={index.toString()}>
                  <Accordion.Header
                    onClick={() => {
                      handleAccordionClick(index.toString());
                    }}>
                    <div className="d-flex w-100">
                      <div className="h3 fw-normal m-0">
                        {serviceData.organisation + " - " + serviceData.date}
                      </div>

                      <div className="flex-grow-1 align-self-center text-end">
                        {activeKey === index.toString() ? (
                          <BsCaretUpFill
                            className="caret_border"
                            size={22}></BsCaretUpFill>
                        ) : (
                          <BsCaretDownFill
                            className="caret_border"
                            size={22}></BsCaretDownFill>
                        )}
                      </div>
                    </div>
                  </Accordion.Header>
                  <Accordion.Body>
                    <Table
                      style={{ borderTop: "1px solid #ccc" }}
                      className="fw-light text-start"
                      responsive="sm">
                      <thead>
                        <tr>
                          <th>Bus ID</th>
                          <th>Route Variant</th>
                          <th>Status</th>
                        </tr>
                      </thead>
                      <tbody>
                        {serviceData.busData.map(
                          (busRunningInfo: BusRunningInfo) => (
                            <tr key={busRunningInfo.busId}>
                              <td>{busRunningInfo.busId}</td>
                              <td>
                                <RouteVariant
                                  formatString={
                                    busRunningInfo.routeVariant
                                  }></RouteVariant>
                              </td>
                              <td>
                                {busRunningInfo.deviationFromTimetable ===
                                  null ||
                                busRunningInfo.deviationFromTimetable === 0 ? (
                                  <span className="text-success">On Time</span>
                                ) : busRunningInfo.deviationFromTimetable >
                                  0 ? (
                                  <span className="text-primary">Late</span>
                                ) : (
                                  <span className="text-danger">Early</span>
                                )}
                              </td>
                            </tr>
                          )
                        )}
                      </tbody>
                    </Table>
                    <ServiceNotes serviceName={serviceData.organisation} />
                  </Accordion.Body>
                </Accordion.Item>
              </Accordion>
            ))}
        </div>
      </div>
    </>
  );
}
export default RunningReport;
