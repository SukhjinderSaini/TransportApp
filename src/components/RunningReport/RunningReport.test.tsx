import React from "react";
import renderer from "react-test-renderer";
import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import fetchMock from "jest-fetch-mock";
import RunningReport from "./RunningReport";
import * as busData from "../../../public/data/bus-services-data.json";

// Mock the fetch function
fetchMock.enableMocks();

describe("RunningReport", () => {
  it("fetches data and renders it correctly", async () => {
    // Mock the response data
    const mockData = busData.data;
    fetchMock.mockResponseOnce(JSON.stringify(mockData));

    render(<RunningReport />);
    await screen.findByText("Bus Reports");
    expect(screen.getByText("Sydney Buses")).toBeInTheDocument();
    expect(screen.getByText("Westbus")).toBeInTheDocument();

    // Add more assertions based on your component's rendering logic
  });

  it("RunningReport renders correctly", () => {
    const tree = renderer.create(<RunningReport />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
