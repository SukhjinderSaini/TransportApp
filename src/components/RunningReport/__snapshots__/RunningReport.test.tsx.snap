// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RunningReport RunningReport renders correctly 1`] = `
<div
  className="d-flex flex-column m-sm-3 m-md-5"
>
  <div
    className="display-6 fw-normal text-start"
  >
    Bus Reports
  </div>
  <div
    className="aling-items-center justify-content-center mt-3"
  />
</div>
`;

exports[`renders correctly 1`] = `
<div
  className="d-flex flex-column m-sm-3 m-md-5"
>
  <div
    className="display-6 fw-normal text-start"
  >
    Bus Reports
  </div>
  <div
    className="aling-items-center justify-content-center mt-3"
  />
</div>
`;

exports[`running report renders correctly 1`] = `
<div
  className="d-flex flex-column m-sm-3 m-md-5"
>
  <div
    className="display-6 fw-normal text-start"
  >
    Bus Reports
  </div>
  <div
    className="aling-items-center justify-content-center mt-3"
  />
</div>
`;
